File: events/2015/14/solution.ts
import { AOCDayResults } from "@repo/utils-ts/returnType";
import { readInputAsArray } from "@repo/utils-ts/readInput";
import { measurePerformance } from "@repo/utils-ts/measurePerformance";

type ReindeerStats = {
  fly_speed: number,
  fly_duration: number,
  rest_duration: number
  points?: number
}

function whosInTheLead(reindeers: Record<string, ReindeerStats>, duration: number): [string[], number] {
  let maxDistance = 0;
  let winningReindeer: string[] = []
  for (const reindeer in reindeers) {    
    let distance = 0;
    let distancePerBurst = reindeers[reindeer].fly_speed * reindeers[reindeer].fly_duration;
    let durationRemaining = duration
    let fullPeriod = reindeers[reindeer].fly_duration + reindeers[reindeer].rest_duration;

    // Can we do a full burst or not?
    if (duration < reindeers[reindeer].fly_duration) {
      distance = reindeers[reindeer].fly_speed * duration
    } else {
      // Handle the initial burst
      distance += distancePerBurst
      durationRemaining -= reindeers[reindeer].fly_duration

      // How many more bursts can we do
      let potentialBursts = Math.floor(durationRemaining / fullPeriod)
      distance += distancePerBurst * potentialBursts

      // Partial Bursts?
      let timeAfterBursts = (durationRemaining - (potentialBursts * fullPeriod))
      if (timeAfterBursts > reindeers[reindeer].rest_duration) {
        let extraTime = timeAfterBursts - reindeers[reindeer].rest_duration
        distance += extraTime * reindeers[reindeer].fly_speed
      }
    }
    
    if (distance > maxDistance) {
      maxDistance = distance
      winningReindeer = [reindeer]
    } else if (distance == maxDistance) {
      winningReindeer.push(reindeer)
    }
  }

  return [winningReindeer, maxDistance]
}

// ----------------------------------------------------------------------------------------------------
// | Part 1
// ----------------------------------------------------------------------------------------------------
function part1(input: string[]): number | undefined {
  let duration = 2503
  let reindeers: Record<string, ReindeerStats> = {}

  // Work everything out for each reindeer
  for (var i = 0; i < input.length; i++) {
    if (input[i].length == 0 ) { continue }
    let parts = input[i].split(" ")
    let stats: ReindeerStats = {
      fly_speed: parseInt(parts[3]),
      fly_duration: parseInt(parts[6]),
      rest_duration: parseInt(parts[13])
    }
    reindeers[parts[0]] = stats
  }

  // Work out overall speed
  let [winningDeer, maxDistance] = whosInTheLead(reindeers, duration)

  return maxDistance
}

// ----------------------------------------------------------------------------------------------------
// | Part 2
// ----------------------------------------------------------------------------------------------------
function part2(input: string[]): number | undefined {
  let duration = 2503
  let reindeers: Record<string, ReindeerStats> = {}

  // Work everything out for each reindeer
  for (var i = 0; i < input.length; i++) {
    if (input[i].length == 0 ) { continue }
    let parts = input[i].split(" ")
    let stats: ReindeerStats = {
      fly_speed: parseInt(parts[3]),
      fly_duration: parseInt(parts[6]),
      rest_duration: parseInt(parts[13]),
      points: 0
    }
    reindeers[parts[0]] = stats
  }

  // Work out the points each second
  for (var s = 1; s < duration+1; s++) {
    // Loop over the reindeer and work it out
    let [winners, maxDistance] = whosInTheLead(reindeers, s)

    for (const index in winners) {
      let winningReindeer = winners[index]
      if (!reindeers[winningReindeer]) {
        continue
      }
      if (reindeers[winningReindeer].points !== undefined) {
        reindeers[winningReindeer].points += 1
      } else {
        reindeers[winningReindeer].points = 1;
      }
    }
  }

  // Who wins?
  let maxPoints = 0
  for (const reindeer in reindeers) {
    if (!reindeers[reindeer].points) { continue }

    if (reindeers[reindeer].points > maxPoints) {
      maxPoints = reindeers[reindeer].points
    }
  }

  return maxPoints
}


// ----------------------------------------------------------------------------------------------------
// | Solve the puzzle
// ----------------------------------------------------------------------------------------------------
function solve() {
  // Get the arguments
  const args = process.argv.slice(2);
  const fileName = args[0] || "input.txt";

  // Start the timer
  const startTimeStamp = new Date(Date.now()).toISOString();
  const startTime = performance.now();

  // Read the input as an array of strings
  const input = readInputAsArray(fileName);

// Run the parts
const p1Result = measurePerformance(() => part1(input));
const p2Result = measurePerformance(() => part2(input));

// End the timer
const endTimeStamp = new Date(Date.now()).toISOString();
const endTime = performance.now();
const duration = endTime - startTime;

// Return the results
return {
  Year: 2015,
  Day: 14,
  Part1: p1Result,
  Part2: p2Result,
  Duration: Math.round(duration * 1000000), // Convert to nanoseconds
  Timestamp: {
    Start: new Date(startTimeStamp).toISOString(),
    End: new Date(endTimeStamp).toISOString(),
  },
} as AOCDayResults;
  
}
const results = solve();
console.log(JSON.stringify(results, null, 2));

File: events/2015/14/solution.rs
use aoc_utils::read_input_as_array;
use aoc_utils::measure_performance;
use aoc_utils::AOCDayResults;
use aoc_utils::AOCTimestamp;
use chrono::{DateTime, Utc};
use std::collections::{HashSet, HashMap};

#[derive(Clone)]
struct ReindeerStats {
  fly_speed: i32,
  fly_duration: i32,
  rest_duration: i32,
  points: i32
}

fn whos_in_the_lead(reindeers: HashMap<String, ReindeerStats>, duration: i32) -> (HashSet<String>, i32) {
  let mut max_distance: i32 = 0;
  let mut winning_reindeer: HashSet<String> = HashSet::new();

  for (reindeer, stats) in reindeers {
    let mut distance: i32 = 0;
    // let stats = reindeers[reindeer];
    let distance_per_burst = stats.fly_speed * stats.fly_duration;
    let mut duration_remaining = duration;
    let full_period = stats.fly_duration + stats.rest_duration;

    if duration < stats.fly_duration {
      distance = stats.fly_speed * duration;
    } else {
      distance += distance_per_burst;
      duration_remaining -= stats.fly_duration;

      let potential_bursts = duration_remaining / full_period;
      distance += distance_per_burst * potential_bursts;

      let time_after_bursts = duration_remaining - (potential_bursts * full_period);
      if time_after_bursts > stats.rest_duration {
        let extra_time = time_after_bursts - stats.rest_duration;
        distance += extra_time * stats.fly_speed;
      }
    }

    if distance > max_distance {
      max_distance = distance;
      winning_reindeer = HashSet::new();
      winning_reindeer.insert(reindeer);
    } else if distance == max_distance {
      winning_reindeer.insert(reindeer);
    }
  }

  return (winning_reindeer, max_distance)
}


// ----------------------------------------------------------------------------------------------------
// | Part 1
// ----------------------------------------------------------------------------------------------------
fn part1(input: &Vec<String>) -> i32 {
  let duration: i32 = 2503;
  let mut reindeers: HashMap<String, ReindeerStats> = HashMap::new();

  for line in input {
    if line.len() == 0 { continue }
    
    let parts: Vec<&str> = line.split(" ").collect();
    let stats = ReindeerStats {
      fly_speed: parts[3].parse::<i32>().unwrap(),
      fly_duration: parts[6].parse::<i32>().unwrap(),
      rest_duration: parts[13].parse::<i32>().unwrap(),
      points: 0
    };
    reindeers.insert(parts[0].to_string(), stats);
  }

  let (_winner, max_distance) = whos_in_the_lead(reindeers, duration);

  max_distance
}

// ----------------------------------------------------------------------------------------------------
// | Part 2
// ----------------------------------------------------------------------------------------------------
fn part2(input: &Vec<String>) -> i32 {
  let duration: i32 = 2503;
  let mut reindeers: HashMap<String, ReindeerStats> = HashMap::new();

  for line in input {
    if line.len() == 0 { continue }
    
    let parts: Vec<&str> = line.split(" ").collect();
    let stats = ReindeerStats {
      fly_speed: parts[3].parse::<i32>().unwrap(),
      fly_duration: parts[6].parse::<i32>().unwrap(),
      rest_duration: parts[13].parse::<i32>().unwrap(),
      points: 0
    };
    reindeers.insert(parts[0].to_string(), stats);
  }

  for s in 1..duration+1{
    let (winners, _distance) = whos_in_the_lead(reindeers.clone(), s);

    for winner in winners {
      if let Some(stats) = reindeers.get_mut(&winner) {
        stats.points += 1;
      }
    }
  }

  let mut max_points = 0;
  for (_reindeer, stats) in reindeers {
    if stats.points > max_points {
      max_points = stats.points;
    }
  }
  
  max_points
}

// ----------------------------------------------------------------------------------------------------
// | Main Function
// ----------------------------------------------------------------------------------------------------
fn main() {
  // Read the arguments
  let args = std::env::args().collect::<Vec<String>>();
  let default_file_name = String::from("input.txt");
  let file_name = args.get(1).unwrap_or(&default_file_name);

  // Start the timer
  let start_timestamp = std::time::SystemTime::now();
  let start_time = std::time::Instant::now();
  
  // Read the input
  let input = read_input_as_array(file_name).unwrap();

  // Run the Parts
  let p1_result = measure_performance(part1, &input);
  let p2_result = measure_performance(part2, &input);

  // Stop the timer
  let end_time = std::time::Instant::now();
  let end_timestamp = std::time::SystemTime::now();
  let duration = end_time.duration_since(start_time).as_nanos();

  let results = AOCDayResults {
    year: 2015,
    day: 14,
    part1: p1_result,
    part2: p2_result,
    duration: duration,
    timestamp: AOCTimestamp {
      start: DateTime::<Utc>::from(start_timestamp).to_rfc3339(),
      end: DateTime::<Utc>::from(end_timestamp).to_rfc3339(),
    },
  };

  let json_results = serde_json::to_string(&results).unwrap();
  println!("{}", json_results);
}